Automattic/pocket-casts-android
# Improve the Google Single Sign-On logic

This change improves the SSO implementation.

- Only the refresh and access tokens are stored in the Android account manager.
- The access and refresh tokens are wrapped in classes `AccessToken` and `RefreshToken` to reduce the complexity of having two similar tokens. (Thanks @mchowning)
- The sync account methods have been moved from the `Settings` class and into `SyncAccountManager`. This new class also wraps the Android `AccountManager` class to simplify dealing with and centralising common logic.
- Hide the change email and change password buttons when logged in using Google.
- Fixes the Sonos linking using tokens.
- Added new server error messages to be localised. 
- Improved the /sync/update call. Removed parameters not required. Moved it into the main Retrofit class.
- Switched the SyncServerRetrofit to use an OkHttp client with no cache.

## Test Steps

### Test create account using a Google account

1. Install a fresh copy of the app
2. Tap 'Continue with Google'
3. Pick a Google account that doesn't already exist in the database (or delete it first)
4. ✅ Verify you are taken to the 'Find your favourite podcasts' page
6. Go to the 'Profile' tab
7. Tap on the email row
8. ✅ Verify the 'Change email address' and 'Change passport' options aren't visible
9. Open the 'Discover' tab
10. Subscribe to a podcast
11. Tap the 'Profile' tab 
12. Tap the 'Refresh now' button

###  Test sign in using a Google account

1. Install a fresh copy of the app, or sign out and clear the app data
2. Tap 'Continue with Google'
3. Pick the Google account which had the Pocket Casts account created for it
4. ✅ Verify the 'Podcasts' tab has the same subscribed to podcast

###  Test create account using an email and password

1. Install a fresh copy of the app, or sign out and clear the app data
2. Tap 'Sign up'
3. Enter an email and password
4.  ✅ Verify you are taken to the 'Find your favourite podcasts' page
6. Go to the 'Profile' tab
7. Tap on the email row
8. ✅ Verify the 'Change email' and 'Change passport' options are visible
9. Open the 'Discover' tab
10. Subscribe to a podcast
11. Tap the 'Profile' tab 
12. Tap the 'Refresh now' button

###  Test sign in using an email and password

1. Install a fresh copy of the app, or sign out and clear the app data
2. Tap 'Log in'
3. Enter the same email and password as the previous test
4. ✅ Verify the 'Podcasts' tab has the same subscribed to podcast

### Test expiring access token 

1. Deploy the staging version of the app to your device
14. Sign in using Google SSO
15. Wait for over an hour (staging access tokens expiry after 1 hour)
16. Open the 'Profile' tab
17. Tap the 'Refresh Now' button
18. ✅ Verify the user is still signed in

### Sonos connection using Google SSO

1. Deploy the release version of the app to your device
19. Login using Google SSO
20. Open the Sonos app
21. Open the settings tab
22. Tap 'Services & Voice'
23. Tap 'Add Service'
24. Tap 'Pocket Casts'
25. Tap 'Add to Sonos'
26. Tap 'Connect to Pocket Casts'
27. ✅ Verify the Pocket Casts app is opened automatically
28. Tap 'Connect'
29. ✅ Verify the Pocket Casts service is successfully added in the Sonos app
